### PR TITLE
How to Prevent Cross Site Scripting in Node.js

### PR DESCRIPTION
**Pull request by Pauline-Mwangi**
### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by Pauline-Mwangi** [comment]
This fixes #1927

**Comment by lalith1403** [comment]
@Pauline-Mwangi Consider adding a headshot pic and short bio under a folder named after you, placed in the author's folder in the eng-ed repository. Check the author's folder for more details.

File: articles/how-to-prevent-cross-site-scripting-in-node.js/index.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+Most developers do not understand how [DOM-based XSS](https://portswigger.net/web-security/cross-site-scripting/dom-based) works.
+
+### How to prevent XSS
+XSS is the leading attack for web applications. Now that we have an understanding of XSS and how it damages applications. The following are techniques used to [prevent XSS vulnerability](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html) in Node.js
+
+#### Output Encoding
+Output encoding translates untrusted input into a safe model. It takes place by running a JavaScript entity encoding in a Node.js platform server prior to sending untrusted data into the browser. 
+
+#### Avoid Inserting Untrusted Data
+Do not pass untrusted data into executable JavaScript functions such as `eval`. In the Node.js `eval` function, enable code injections. The `eval` function takes a string argument and interprets it as a JavaScript code allowing the attacker to execute it.
```
Comment (lalith1403) [inline articles/how-to-prevent-cross-site-scripting-in-node.js/index.md:81]: This part is not clear. What data and what is the role of the function `eval`? What are code injections? Explain it in the proper order.

File: articles/how-to-prevent-cross-site-scripting-in-node.js/index.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+The only secure way to use if you insert untrusted data into the JavaScript code is inside a quoted (")data value. That is the proper context for inserting data into a JavaScript code. For example,
```
Comment (lalith1403) [inline articles/how-to-prevent-cross-site-scripting-in-node.js/index.md:83]: This sentence is complex. Break it down into simpler sentences.

File: articles/how-to-prevent-cross-site-scripting-in-node.js/index.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+It is essential to carry out server-side input validation when preventing XSS attacks. Server-side is mainly preferred for input validation because it can easily be bypassed when done via the client-side.
+
+While validating user input with server-side JavaScript (node.js), try the npm module `strip-js`. It does the following:
+- Sanitizes data
+- Do away with script tags
+- Remove attributes that contain JavaScript code, i.e., `onclick` and `onerror`.
+- Remove “href” attributes with JavaScript code.
+
+#### Sanitizing input 
+Properly sanitized input ensures that users can submit only trusted data. It ensures that only expected data is inputted, i.e. (only authorizing input fields to accept parenthesis, numbers, and hyphens).  Sanitizing data in an HTTP request before execution ensures data is validated, escaped, and filtered before it is outputted to the user.
```
Comment (lalith1403) [inline articles/how-to-prevent-cross-site-scripting-in-node.js/index.md:108]: What is sanitized input? Explain a bit more in this section.

File: articles/how-to-prevent-cross-site-scripting-in-node.js/index.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+Another way to prevent XSS is preventing cookies from browser JS scripts. This is achieved by setting the cookie `httpOnly` value to `true` in the below code.
```
Comment (lalith1403) [inline articles/how-to-prevent-cross-site-scripting-in-node.js/index.md:110]: Does this come under sanitizing input section, or is it a different one? It is not clear. Also, explain what happens when the cookie value is set to `True.`

**Review by lalith1403 — CHANGES_REQUESTED** [review]
Well-written article. The initial part of the article is crisp and easy to understand. Well done. I have added a few modifications that would enhance the article further in terms of readability. Review them and make the changes accordingly.

**Comment by Pauline-Mwangi** [comment]
Thank you @lalith1403 I made the changes you requested

**Comment by hectorkambow** [comment]
plagiarism check done

**Comment by lalith1403** [comment]
The folder name is `how-to-prevent-cross-site-scripting-in-node.js`. I suggest you change it to `how-to-prevent-cross-site-scripting-in-node-js` to avoid errors in the URL.

**Comment by Pauline-Mwangi** [comment]
@lalith1403 I made the final change

**Comment by hectorkambow** [comment]
just need to preview and published - when repo changes are finalized

NOTE: Thank you - we are making some changes to the main repo now - and this is delaying a few article from being published as we want to implement the changes 1st - once these are done - we will publish this one asap. Thank you 🚀

**Comment by hectorkambow** [comment]
published a couple days ago - had to create a new PR to prevent merge issues 👍 
https://www.section.io/engineering-education/how-to-prevent-cross-site-scripting-in-node-js/

**Comment by hectorkambow** [comment]
FYI please be sure to do a git pull on your forked repo before submitting your next PR to avoid any potential merge issues.

Thank you v much again for your patience as we worked the new repo file structure.

**Pull request closed without merging**